src/store/Workspaces/cheWorkspaces/index.ts

```diff
@@ -266,6 +275,39 @@ export const actionCreators: ActionCreators = {
     }
   },
 
+  restartWorkspace: (workspace: che.Workspace): AppThunk<KnownAction, Promise<void>> => async (dispatch, getState): Promise<void> => {
+    const defer: IDeferred<void> = getDefer();
+    const toDispose = new DisposableCollection();
+    const onStatusChangeCallback = async (status: string) => {
+      if (status !== WorkspaceStatus.STOPPED && status !== WorkspaceStatus.ERROR) {
+        return;
+      }
+      try {
+        await dispatch(actionCreators.startWorkspace(workspace));
+        defer.resolve();
+      } catch (e) {
+        defer.reject(e);
+      }
+      toDispose.dispose();
+    };
+    if (workspace.status === WorkspaceStatus.STOPPED || workspace.status === WorkspaceStatus.ERROR) {
+      await onStatusChangeCallback(workspace.status);
+    } else {
+      subscribeToStatusChange(workspace, dispatch);
+      try {
+        await dispatch(actionCreators.stopWorkspace(workspace));
+      } catch (error) {
+        defer.reject(error);
+      }
+      const workspaceId = workspace.id;
+      onStatusChangeCallbacks.set(workspaceId, onStatusChangeCallback);
+      toDispose.push({
+        dispose: () => onStatusChangeCallbacks.delete(workspaceId)
+      });
+    }
+    return defer.promise;
```

Comment:
I suppose here is the best place to create an error message if the restart fails.

Reply:
I have fixed it.